src/crimson/osd/osd_operation.h

```diff
@@ -0,0 +1,233 @@
+protected:
+  virtual void dump_detail(Formatter *f) const = 0;
+
+public:
+  template <typename... T>
+  blocking_future<T...> make_blocking_future(seastar::future<T...> &&f) {
+    return blocking_future(this, std::move(f));
+  }
+
+  void dump(Formatter *f) const;
+
+  virtual const char *get_type_name() const = 0;
+
+  virtual ~Blocker() = default;
+};
+
+template <typename T>
+class BlockerT : public Blocker {
+public:
+  const char *get_type_name() const final {
+    return T::type_name;
+  }
+
+  virtual ~BlockerT() = default;
+};
+
+class Operation : public boost::intrusive_ref_counter<
+  Operation, boost::thread_unsafe_counter> {
+  friend class OperationRegistry;
+  registry_hook_t registry_hook;
+
+  std::vector<Blocker*> blockers;
+  uint64_t id = 0;
+  void set_id(uint64_t in_id) {
+    id = in_id;
+  }
+protected:
+  virtual void dump_detail(Formatter *f) const = 0;
+
+public:
+  uint64_t get_id() const {
+    return id;
+  }
+
+  virtual OperationTypeCode get_type() const = 0;
+  virtual const char *get_type_name() const = 0;
+  virtual void print(std::ostream &) const = 0;
+
+  void add_blocker(Blocker *b) {
+    blockers.push_back(b);
+  }
+
+  void clear_blocker(Blocker *b) {
+    auto iter = std::find(blockers.begin(), blockers.end(), b);
+    if (iter != blockers.end()) {
+      blockers.erase(iter);
+    }
+  }
+
+  template <typename... T>
+  seastar::future<T...> with_blocking_future(blocking_future<T...> &&f) {
+    if (f.fut.available() || f.fut.failed()) {
+      return std::move(f.fut);
+    }
+    ceph_assert(f.blocker);
+    add_blocker(f.blocker);
+    return std::move(f.fut).then_wrapped([this, blocker=f.blocker](auto &&arg) {
+      clear_blocker(blocker);
+      return std::move(arg);
+    });
+  }
+
+  void dump(Formatter *f);
+  void dump_brief(Formatter *f);
+  virtual ~Operation() = default;
+};
+using OperationRef = boost::intrusive_ptr<Operation>;
+
+std::ostream &operator<<(std::ostream &, const Operation &op);
+
+template <typename T>
+class OperationT : public Operation {
+
+protected:
+  virtual void dump_detail(Formatter *f) const = 0;
+
+public:
+  static constexpr const char *type_name = OP_NAMES[static_cast<int>(T::type)];
+  using IRef = boost::intrusive_ptr<T>;
+
+  OperationTypeCode get_type() const final {
+    return T::type;
+  }
+
+  const char *get_type_name() const final {
+    return T::type_name;
+  }
+
+  virtual ~OperationT() = default;
+};
+
+class OperationRegistry {
+  friend class Operation;
+  using op_list_member_option = boost::intrusive::member_hook<
+    Operation,
+    registry_hook_t,
+    &Operation::registry_hook
+    >;
+  using op_list = boost::intrusive::list<
+    Operation,
+    op_list_member_option,
+    boost::intrusive::constant_time_size<false>>;
+
+  std::array<
+    op_list,
+    static_cast<int>(OperationTypeCode::last_op)
+  > registries;
+
+  std::array<
+    uint64_t,
+    static_cast<int>(OperationTypeCode::last_op)
+  > op_id_counters = {0};
```

Comment:
should be:
op_id_counters;

The ={0} is misleading. The array is value-init to 0 as a member variable. 
(i.e. using ={17} will not make all elements equal 17).

See https://www.godbolt.org/z/B3FFfj

Reply:
Or  ' = {}'

Reply:
Seems it needs to be = {} for value initialization, thanks!